In [1]:
import pyspark
from pyspark.sql import SparkSession
from pyspark.sql import types

In [42]:
spark = SparkSession.builder \
    .master("local[*]") \
    .appName('test') \
    .getOrCreate()

24/03/04 20:14:03 WARN SparkSession: Using an existing Spark session; only runtime SQL configurations will take effect.


### Question 1. Spark version

In [43]:
spark.version

'3.3.2'

### Question 2. FHV October 2019 partition size

Read the October 2019 FHV into a Spark Dataframe with a schema as we did in the lessons.

Repartition the Dataframe to 6 partitions and save it to parquet.

What is the average size of the Parquet (ending with .parquet extension) Files that were created (in MB)? Select the answer which most closely matches.
- 1MB
- **6MB**
- 25MB
- 87MB

In [86]:
!wget https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhv/fhv_tripdata_2019-10.csv.gz

--2024-03-04 20:25:09--  https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhv/fhv_tripdata_2019-10.csv.gz
Resolving github.com (github.com)... 140.82.121.3
Connecting to github.com (github.com)|140.82.121.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/efdfcf82-6d5c-44d1-a138-4e8ea3c3a3b6?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAVCODYLSA53PQK4ZA%2F20240304%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240304T202509Z&X-Amz-Expires=300&X-Amz-Signature=0ef79967bb8add4912899e1091f1103f656af6c541eb9391d39adbe9efb50907&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=513814948&response-content-disposition=attachment%3B%20filename%3Dfhv_tripdata_2019-10.csv.gz&response-content-type=application%2Foctet-stream [following]
--2024-03-04 20:25:10--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/efdfcf82-6d5c-

In [88]:
import pandas as pd
data = pd.read_csv('fhv_tripdata_2019-10.csv.gz',compression='gzip')
print(data)

        dispatching_base_num      pickup_datetime     dropOff_datetime  \
0                     B00009  2019-10-01 00:23:00  2019-10-01 00:35:00   
1                     B00013  2019-10-01 00:11:29  2019-10-01 00:13:22   
2                     B00014  2019-10-01 00:11:43  2019-10-01 00:37:20   
3                     B00014  2019-10-01 00:56:29  2019-10-01 00:57:47   
4                     B00014  2019-10-01 00:23:09  2019-10-01 00:28:27   
...                      ...                  ...                  ...   
1897488               B03160  2019-10-31 23:38:00  2019-10-31 23:48:00   
1897489               B03160  2019-10-31 23:11:00  2019-10-31 23:43:00   
1897490               B03160  2019-10-31 23:13:00  2019-10-31 23:41:00   
1897491               B03186  2019-10-31 23:02:32  2019-10-31 23:09:53   
1897492               B03186  2019-10-31 23:19:24  2019-10-31 23:24:31   

         PUlocationID  DOlocationID  SR_Flag Affiliated_base_number  
0               264.0         264.0      

In [113]:
schema = types.StructType([
    types.StructField('dispatching_base_num', types.StringType(), True),
    types.StructField('pickup_datetime', types.TimestampType(), True),
    types.StructField('dropoff_datetime', types.TimestampType(), True),
    types.StructField('PULocationID', types.IntegerType(), True),
    types.StructField('DOLocationID', types.IntegerType(), True),
    types.StructField('SR_Flag', types.StringType(), True),
    types.StructField('Affiliated_base_number', types.StringType(), True),
])

In [114]:
df = spark.read \
    .option("header", "true") \
    .schema(schema) \
    .csv('fhv_tripdata_2019-10.csv.gz', sep=',')

df = df.repartition(6)

df.write.parquet('fhv/2019/10')

In [115]:
!ls -lh fhv/2019/10

total 39M
-rw-r--r-- 1 olga olga    0 Mar  4 20:38 _SUCCESS
-rw-r--r-- 1 olga olga 6.4M Mar  4 20:38 part-00000-5aec6f27-98a0-46dc-935c-382b02350165-c000.snappy.parquet
-rw-r--r-- 1 olga olga 6.4M Mar  4 20:38 part-00001-5aec6f27-98a0-46dc-935c-382b02350165-c000.snappy.parquet
-rw-r--r-- 1 olga olga 6.4M Mar  4 20:38 part-00002-5aec6f27-98a0-46dc-935c-382b02350165-c000.snappy.parquet
-rw-r--r-- 1 olga olga 6.4M Mar  4 20:38 part-00003-5aec6f27-98a0-46dc-935c-382b02350165-c000.snappy.parquet
-rw-r--r-- 1 olga olga 6.4M Mar  4 20:38 part-00004-5aec6f27-98a0-46dc-935c-382b02350165-c000.snappy.parquet
-rw-r--r-- 1 olga olga 6.4M Mar  4 20:38 part-00005-5aec6f27-98a0-46dc-935c-382b02350165-c000.snappy.parquet


In [139]:
df = spark.read.parquet('fhv/2019/10')

In [140]:
df.printSchema()

root
 |-- dispatching_base_num: string (nullable = true)
 |-- pickup_datetime: timestamp (nullable = true)
 |-- dropoff_datetime: timestamp (nullable = true)
 |-- PULocationID: integer (nullable = true)
 |-- DOLocationID: integer (nullable = true)
 |-- SR_Flag: string (nullable = true)
 |-- Affiliated_base_number: string (nullable = true)



In [141]:
df.show()

+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|dispatching_base_num|    pickup_datetime|   dropoff_datetime|PULocationID|DOLocationID|SR_Flag|Affiliated_base_number|
+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|              B02784|2019-10-01 09:55:38|2019-10-01 10:05:43|          89|          85|   null|                  null|
|              B02429|2019-10-21 04:15:47|2019-10-21 04:36:04|         264|         264|   null|                B02429|
|              B01482|2019-10-19 12:00:00|2019-10-19 12:20:00|         264|         264|   null|                B01482|
|              B03015|2019-10-11 14:28:00|2019-10-11 14:32:44|         264|         216|   null|                B03015|
|              B01529|2019-10-21 18:00:26|2019-10-21 18:07:21|         264|          80|   null|                B01529|
|              B00477|2019-10-03 19:30:3

### Question 3. Count records on 15th of October
Count records

How many taxi trips were there on the 15th of October?

Consider only trips that started on the 15th of October.
- 108,164
- 12,856
- 452,470
- **62,610**

In [119]:
from pyspark.sql import functions as F

In [120]:
df \
    .withColumn('pickup_date', F.to_date(df.pickup_datetime)) \
    .filter("pickup_date = '2019-10-15'") \
    .count()

62610

In [121]:
df.registerTempTable('fhv_2019_10')

In [122]:
spark.sql("""
SELECT
    COUNT(1)
FROM 
    fhv_2019_10
WHERE
    to_date(pickup_datetime) = '2019-10-15';
""").show()

+--------+
|count(1)|
+--------+
|   62610|
+--------+



### Question 4. The longest trip

Longest trip for each day

What is the length of the longest trip in the dataset in hours?
- **631,152.50 Hours**
- 243.44 Hours
- 7.68 Hours
- 3.32 Hours

In [123]:
df.columns

['dispatching_base_num',
 'pickup_datetime',
 'dropoff_datetime',
 'PULocationID',
 'DOLocationID',
 'SR_Flag',
 'Affiliated_base_number']

In [128]:
# Calculate Time difference in Hours
hour_dur=df.withColumn('pickup_datetime',F.unix_timestamp(F.col('pickup_datetime')))\
  .withColumn('dropoff_datetime',F.unix_timestamp(F.col('dropoff_datetime')))\
  .withColumn('diff_in_hours',(F.col("dropoff_datetime") - F.col('pickup_datetime'))/3600)

In [129]:
from pyspark.sql.functions import max

In [130]:
hour_dur.select(max(hour_dur.diff_in_hours)).show()

+------------------+
|max(diff_in_hours)|
+------------------+
|          631152.5|
+------------------+



### Question 5. Spark UI port

User Interface

Spark’s User Interface which shows the application's dashboard runs on which local port?

- 80
- 443
- **4040**
- 8080

###  Question 6. Least frequent pickup location zone
Least frequent pickup location zone

Using the zone lookup data and the FHV October 2019 data, what is the name of the LEAST frequent pickup location Zone?

- East Chelsea
- **Jamaica Bay**
- Union Sq
- Crown Heights North

In [132]:
!wget https://s3.amazonaws.com/nyc-tlc/misc/taxi+_zone_lookup.csv

--2024-03-04 20:46:14--  https://s3.amazonaws.com/nyc-tlc/misc/taxi+_zone_lookup.csv
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.217.117.112, 52.216.213.120, 52.216.52.0, ...
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.217.117.112|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 12322 (12K) [application/octet-stream]
Saving to: ‘taxi+_zone_lookup.csv’

taxi+_zone_lookup.c 100%[===================>]  12.03K  --.-KB/s    in 0s      

2024-03-04 20:46:14 (110 MB/s) - ‘taxi+_zone_lookup.csv’ saved [12322/12322]



In [150]:
df_zones = spark.read \
    .option("header", "true") \
    .csv('taxi+_zone_lookup.csv')

In [151]:
df_zones.columns

['LocationID', 'Borough', 'Zone', 'service_zone']

In [145]:
df.columns

['dispatching_base_num',
 'pickup_datetime',
 'dropoff_datetime',
 'PULocationID',
 'DOLocationID',
 'SR_Flag',
 'Affiliated_base_number']

In [152]:
df_zones.createOrReplaceTempView('zones')

In [155]:
spark.sql("""
SELECT
    zones.Zone,
    COUNT(*) cnt
FROM 
    fhv_2019_10,zones 
where fhv_2019_10.DOLocationID = zones.LocationID
GROUP BY zones.Zone
ORDER BY cnt ASC
""").show()

+--------------------+---+
|                Zone|cnt|
+--------------------+---+
|Governor's Island...|  4|
|       Rikers Island|  4|
|         Jamaica Bay| 14|
|        Battery Park| 35|
|       Broad Channel| 36|
|Breezy Point/Fort...| 54|
|        Astoria Park| 73|
|     Freshkills Park|132|
|Saint Michaels Ce...|162|
| Green-Wood Cemetery|261|
|       Willets Point|301|
|     Columbia Street|321|
|Marine Park/Floyd...|339|
|    Roosevelt Island|348|
|         City Island|389|
|     Cambria Heights|404|
|     Randalls Island|404|
|       Rockaway Park|412|
|  Brooklyn Navy Yard|455|
|Stuy Town/Peter C...|466|
+--------------------+---+
only showing top 20 rows

